feat: support backoff/retry in OTLP

scottgerring opened the pull request:
Fixes #3081, building on the work started by @AaronRM 🤝  

## Changes

### A new **retry** module added to `opentelemetry-sdk` 
Models the sorts of retry an operation may request (retry / can't retry / throttle), and provides a helper `retry_with_backoff` mechanism that can be used to wrap up a retryable operation and retry it. The helper relies on `experimental_async_runtime` for its runtime abstraction, to provide the actual pausing. It also takes a lambda to classify the error, so the caller can inform the retry mechanism if a retry is required.

### A new **retry_classification** module added to `opentelemetry-otlp`
This bit takes the actual error responses that we get back over OTLP and maps them back to the retry model. Because this is OTLP-specific stuff it belongs here rather than alongside the retry code. 

### Retry binding
... happens in each one of the concrete exporters to tie it all together. 

Also ...

* Extended exporter builders to allow the user to customise default retry policy
* Added new feature flags experimental-http-retry and experimental-grpc-retry which pull in the experimental-async-runtime dep and set everything up. This way we can get going with this _now_ without having to stabilise the experimental-async-runtime feature.

## Open Questions



## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

scottgerring commented on opentelemetry-otlp/src/exporter/tonic/logs.rs:
```diff
@@ -58,40 +63,65 @@ impl TonicLogsClient {
-                otel_debug!(name: "TonicLogsClient.ExportFailed", error = &error);
-                Err(OTelSdkError::InternalFailure(error))
-            }
+        let batch = Arc::new(batch);
```
see if we can avoid this

scottgerring commented on opentelemetry-otlp/Cargo.toml:
```diff
@@ -35,6 +35,7 @@ tracing = {workspace = true, optional = true}
 
 prost = { workspace = true, optional = true }
 tonic = { workspace = true, optional = true }
+tonic-types = { workspace = true, optional = true }
```
Needed for gRPC error type

scottgerring commented on opentelemetry-otlp/src/exporter/http/logs.rs:
```diff
@@ -4,9 +4,124 @@ use http::{header::CONTENT_TYPE, Method};
+        .map_err(|e| OTelSdkError::InternalFailure(e.message))
+    }
+
+    #[cfg(not(feature = "http-retry"))]
```
This is a massive duplication of code.

If we decide that http export always has retry behaviour, which means we include the unstable runtime feature, then we can remove all of this.

Alternatively we can provide `export` once, with all the extra pomp and fanfare to support retry, and then just use the stub "don't actually retry" impl. There would be some slight runtime overhead to this, but the codebase would be much simpler.

scottgerring replied:
Comment above applies to the other HTTP exporters also.

scottgerring replied:
I've also not touched this for the tonic exporters yet, but once we have the HTTP ones in shape I can continue the pattern over there.

scottgerring replied:
Because of the way the tonic exporters are structured its harder to cleanly pull it apart. I've struck what I think is a sensible compromise over there, adding a wrapper for the retry function that subs in retry/no-op based on feature flags. I think this is better than going too wild with traits/generics/lambdas to squeeze the last little bit of code re-use out in terms of clarity.

bantonsson commented on opentelemetry-otlp/src/exporter/http/logs.rs:
```diff
@@ -4,9 +4,124 @@ use http::{header::CONTENT_TYPE, Method};
+                let client = self
+                    .client
+                    .lock()
+                    .map_err(|e| HttpExportError {
```
I think the code would be easier to read if there were two constructors for `HttpExportError`. Something like `HttpExportError::new(500, format!("Mutex lock failed: {e}"))` and `HttpExportError::with_retry_after(...)`. It would cut down on the boiler plate.

scottgerring replied:
Good catch - have implemented

bantonsson commented on opentelemetry-otlp/src/exporter/http/logs.rs:
```diff
@@ -4,9 +4,124 @@ use http::{header::CONTENT_TYPE, Method};
 impl LogExporter for OtlpHttpClient {
+    #[cfg(feature = "http-retry")]
+    async fn export(&self, batch: LogBatch<'_>) -> OTelSdkResult {
+        let policy = RetryPolicy {
```
I think a default retry policy makes sense.

bantonsson reviewed (approved):
👍🏼 for the HTTP code. I can't see a clear way to reuse more of the Tonic code.

scottgerring commented on opentelemetry-sdk/src/retry.rs:
```diff
@@ -0,0 +1,461 @@
+fn generate_jitter(max_jitter: u64) -> u64 {
+    let now = SystemTime::now();
+    let nanos = now
+        .duration_since(SystemTime::UNIX_EPOCH)
```
I would expect hashed nanoseconds should be sufficient ? Likelihood of getting the same ns twice at once negligible.

bantonsson replied:
Indeed. The risk of getting the same ns twice is negligible.

lalitb replied:
Makes sense. I was probably thinking of using a proper RNG, and also SystemTime syscall perf - but the current implementation works fine for now. We can revisit if needed, not a blocker.​​​​​​​​​​​​​​​​

scottgerring commented on opentelemetry-sdk/src/lib.rs:
```diff
@@ -167,3 +167,6 @@ impl<T> From<std::sync::PoisonError<T>> for InMemoryExporterError {
 }
+
+/// Retry logic for exporting telemetry data.
+pub mod retry;
```
I would expect any exporter would be interested in it, but I appreciate your point about stretching the API. I can move it 👌

lalitb replied:
My thoughts are: if only OTLP uses this retry logic and the other core exporters in main repo don’t use it, we should keep it scoped to the OTLP module. Will leave it to you to decide :)

scottgerring commented:
@lalitb, i've moved the retry down into otlp (rather than sdk), implemented the date parsing, and added a hash around the nanos for jitter. I think this should be ready to go! Would you like to give it another skim?

lalitb commented on opentelemetry-otlp/src/exporter/http/mod.rs:
```diff
@@ -298,16 +363,148 @@ impl HttpExporterBuilder {
+        #[cfg(feature = "experimental-http-retry")]
+        {
+            use crate::retry::retry_with_backoff;
+            use opentelemetry_sdk::runtime::Tokio;
```
Sorry if this was covered earlier — this change introduces a `Tokio` dependency for setups using the default (thread-based) BatchProcessor combined with `reqwest-blocking`, which otherwise wouldn’t require the runtime dependency. 
Not a blocker since it’s under a feature flag, but just wanted to confirm if this behavior is expected.

lalitb reviewed (approved):
LGTM overall - implementation looks solid and well-structured. I left a comment on the Tokio dependency; not a blocker, but something we can revisit or refine once the feature flag is removed.

cijothomas commented on opentelemetry-otlp/src/retry.rs:
```diff
@@ -0,0 +1,495 @@
+
+                match error_type {
+                    RetryErrorType::NonRetryable => {
+                        otel_warn!(name: "OtlpRetry", message = format!("Operation {:?} failed with non-retryable error: {:?}", operation_name, err));
```
Let's stick with structured logging as much as possible instead of stringifying. i.e Operation and Error should be own fields.

cijothomas commented on opentelemetry-otlp/src/retry.rs:
```diff
@@ -0,0 +1,495 @@
+                    RetryErrorType::Throttled(server_delay) if attempt < policy.max_retries => {
+                        attempt += 1;
+                        // Use server-specified delay (overrides exponential backoff)
+                        otel_warn!(name: "OtlpRetry", message = format!("Retrying operation {:?} after server-specified throttling delay: {:?}", operation_name, server_delay));
```
please downgrade to info level.

cijothomas commented on opentelemetry-otlp/CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ Released 2025-Sep-25
 
 - Update `opentelemetry-proto` and `opentelemetry-http` dependency version to 0.31.0
 - Add HTTP compression support with `gzip-http` and `zstd-http` feature flags
+- Add retry with exponential backoff and throttling support for HTTP and gRPC exporters
```
lets add some more details here so the user reading the changelog would know how to use this feature. (given this is experimental and opt-in).
